tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2735,6 +2735,34 @@ class HLSLExternalSource : public ExternalSemaSource {
 
   HRESULT CombineDimensions(QualType leftType, QualType rightType, QualType *resultType);
 
+  clang::TypedefDecl *HLSLExternalSource::GetMatrixShorthandTypes(
+      QualType type, HLSLScalarType scalarType, UINT rowCount, UINT colCount) {
+    DXASSERT_NOMSG(scalarType != HLSLScalarType::HLSLScalarType_unknown &&
+                   rowCount >= 0 && rowCount <= 4 && colCount >= 0 &&
+                   colCount <= 4);
+    TypedefDecl *qts =
+        m_matrixShorthandTypes[scalarType][rowCount - 1][colCount - 1];
+    if (qts == nullptr) {
+      qts = CreateMatrixSpecializationShorthand(*m_context, type, scalarType,
+                                                rowCount, colCount);
+      m_matrixShorthandTypes[scalarType][rowCount - 1][colCount - 1] = qts;
+    }
+    return qts;
+  }
+
+  clang::TypedefDecl *HLSLExternalSource::GetVectorShorthandType(
+      QualType type, HLSLScalarType scalarType, UINT colCount) {
```

Comment:
These methods should incorporate Lookup[Vector|Matrix]Type, eliminating the need for the QualType parameter supplied by the caller.  Further, this call is only necessary when creating the specialization shorthand, so it can be placed within the if (qts == nullptr) block.